contracts/Staking.sol

```diff
@@ -0,0 +1,487 @@
+     * @return Amount of tokens staked but not locked by given account
+     */
+    function unlockedBalanceOf(address _account) public view returns (uint256) {
+        uint256 unlockedTokens = totalStakedFor(_account);
+
+        Account storage account = accounts[_account];
+        for (uint256 i = 0; i < account.activeLockIds.length; i++) {
+            unlockedTokens = unlockedTokens.sub(account.locks[account.activeLockIds[i]].amount);
+        }
+
+        return unlockedTokens;
+    }
+
+    /**
+     * @notice Check if `_account`'s lock #`_lockId` can be unlocked
+     * @param _account Owner of lock
+     * @param _lockId Id of the lock for the given account
+     * @return Whether given lock of given account can be unlocked
+     */
+    function canUnlock(address _account, uint256 _lockId) public view returns (bool) {
+        Lock storage _lock = accounts[_account].locks[_lockId];
+
+        if (msg.sender == address(_lock.manager) ||
+            (msg.sender == _account && _lock.manager.canUnlock(_account, _lockId, _lock.data))) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /*
+    function multicall(bytes[] _calls) public {
+        for(uint i = 0; i < _calls.length; i++) {
+            require(address(this).delegatecall(_calls[i]), ERROR_MULTICALL_DELEGATECALL);
+        }
+    }
+    */
+
+    /* Internal functions */
+
+    function _stakeFor(address _account, uint256 _amount, bytes _data) internal {
+        // stake 0 tokens makes no sense
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // transfer tokens
+        require(stakingToken.transferFrom(msg.sender, this, _amount), ERROR_TOKEN_TRANSFER);
+
+        // process Stake
+        _modifyStakeBalance(_account, _amount, true);
+
+        // Update global history
+        _updateTotalStaked();
+
+        emit Staked(_account, _amount, totalStakedFor(_account), _data);
+    }
+
+    function _modifyStakeBalance(address _account, uint256 _by, bool _increase) internal {
+        uint256 currentStake = totalStakedFor(_account);
+
+        uint256 newStake;
+        if (_increase) {
+            newStake = currentStake.add(_by);
+        } else {
+            newStake = currentStake.sub(_by);
+        }
+
+        _setStakedFor(_account, newStake);
+    }
+
+    function _updateTotalStaked() internal {
+        totalStakedHistory.add(getBlockNumber64(), stakingToken.balanceOf(this));
+    }
+
+    function _setStakedFor(address _account, uint256 _amount) internal {
+        stakedHistory[_account].add(getBlockNumber64(), _amount);
+    }
+
+    function _unlock(address _account, uint256 _lockId) internal {
+        Account storage account = accounts[_account];
+        Lock storage _lock = account.locks[_lockId];
+
+        _lock.unlockedAt = getTimestamp64();
+
+        // remove from active locks, replacing it by the last one in the array
+        for (uint256 i = 0; i < account.activeLockIds.length; i++) {
```

Comment:
Cache the length in a variable to avoid solidity potentially doing duplicated `SLOAD`s